### PR TITLE
`get_last_with_len`: lint `VecDeque` and any deref to slice

### PR DESCRIPTION
**Pull request by Alexendoo**
changelog: [`get_last_with_len`]: lint `VecDeque` and any deref to slice

Previously only `Vec`s were linted, this will now catch any usages on slices, arrays, etc. It also suggests `.back()` for `VecDeque`s

Also moves the lint into `methods/`

### CODE REVIEW

File: clippy_lints/src/methods/get_last_with_len.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+use rustc_hir::{BinOpKind, Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty;
+use rustc_span::source_map::Spanned;
+use rustc_span::sym;
+
+use super::GET_LAST_WITH_LEN;
+
+pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, recv: &Expr<'_>, arg: &Expr<'_>) {
+    if let Some(recv_local) = path_to_local(recv)
```
Comment (Jarcho) [inline clippy_lints/src/methods/get_last_with_len.rs:15]: This is significantly more restrictive than it used to be. Something like `x.y.get(..)` or `x[0].get(..)` would have been linted before.

I'd think going back to using `SpanlessEq`, but adding a call to `Expr::can_have_side_effects` would be better.

File: clippy_lints/src/methods/get_last_with_len.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        && lhs_path.ident.name == sym::len
+
+        // RHS of subtraction is 1
+        && let ExprKind::Lit(rhs_lit) = &rhs.kind
+        && let LitKind::Int(1, ..) = rhs_lit.node
+
+        // check that recv == lhs_recv `recv.get(lhs_recv.len() - 1)`
+        && path_to_local(lhs_recv) == Some(recv_local)
+    {
+        let method = match cx.typeck_results().expr_ty_adjusted(recv).peel_refs().kind() {
```
Comment (Jarcho) [inline clippy_lints/src/methods/get_last_with_len.rs:36]: Quick question here. This allows any type which derefs to a `VecDeque` or a slice. Is that intentional?

Comment (Alexendoo) [reply clippy_lints/src/methods/get_last_with_len.rs:36]: It is yeah

**Comment by Jarcho** [comment]
Can you add a test for a field access as well just to make sure it doesn't break in the future.

Otherwise looks good. @xFrednet I don't know if you want to add anything before merging.

File: clippy_lints/src/methods/get_last_with_len.rs (hunk 2 above)
Comment (Jarcho) [reply clippy_lints/src/methods/get_last_with_len.rs:36]: I don't know if you want to handle this horrible case or not:
```rust
struct Foo(Vec<u32>);
impl Deref for Foo { type Target = Vec<u32>; ... }
impl Foo {
    // Only replace `last`, but not `get`
    fn last(&self) -> SomeType { ... }
}
```

I don't think it's necessary to do so. It's kind of a pathological case.

**Comment by Alexendoo** [comment]
👍 added

File: clippy_lints/src/methods/get_last_with_len.rs (hunk 2 above)
Comment (Alexendoo) [reply clippy_lints/src/methods/get_last_with_len.rs:36]: I'm happy to leave that one out, can't imagine it coming up

**Review by Jarcho — APPROVED** [review]

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
I'm a big fan of this change, nice cleanup!!! :muscle: 

Thank you, @Jarcho, for the review.

@bors r=Jarcho,xFrednet

**Comment by xFrednet** [comment]
Weird, seems like some CI weirdness. Maybe bors wanted a break ^^

@bors retry

**Pull request merged**